model/utils.go

```diff
@@ -575,73 +512,25 @@ func IsValidId(value string) bool {
 	return true
 }
 
-// Copied from https://golang.org/src/net/dnsclient.go#L119
-func IsDomainName(s string) bool {
-	// See RFC 1035, RFC 3696.
-	// Presentation format has dots before every label except the first, and the
-	// terminal empty label is optional here because we assume fully-qualified
-	// (absolute) input. We must therefore reserve space for the first and last
-	// labels' length octets in wire format, where they are necessary and the
-	// maximum total length is 255.
-	// So our _effective_ maximum is 253, but 254 is not rejected if the last
-	// character is a dot.
-	l := len(s)
-	if l == 0 || l > 254 || l == 254 && s[l-1] != '.' {
-		return false
-	}
-
-	last := byte('.')
-	ok := false // Ok once we've seen a letter.
-	partlen := 0
-	for i := 0; i < len(s); i++ {
-		c := s[i]
-		switch {
-		default:
-			return false
-		case 'a' <= c && c <= 'z' || 'A' <= c && c <= 'Z' || c == '_':
-			ok = true
-			partlen++
-		case '0' <= c && c <= '9':
-			// fine
-			partlen++
-		case c == '-':
-			// Byte before dash cannot be dot.
-			if last == '.' {
-				return false
-			}
-			partlen++
-		case c == '.':
-			// Byte before dot cannot be dot, dash.
-			if last == '.' || last == '-' {
-				return false
-			}
-			if partlen > 63 || partlen == 0 {
-				return false
-			}
-			partlen = 0
-		}
-		last = c
-	}
-	if last == '-' || partlen > 63 {
-		return false
-	}
-
-	return ok
-}
-
 func RemoveDuplicateStrings(in []string) []string {
-	out := []string{}
-	seen := make(map[string]bool, len(in))
-
-	for _, item := range in {
-		if !seen[item] {
-			out = append(out, item)
-
-			seen[item] = true
+	// In-place de-dup.
+	// Copied from https://github.com/golang/go/wiki/SliceTricks#in-place-deduplicate-comparable
```

Comment:
I am all for optimizations but unless we start dealing with thousands of strings at a time does it really matter? Aside from the added complexity (which is not that bad anyway) what slightly worries me is the change in input/output access. Before we were treating the input as immutable and returning a new copy of the data. I think that's usually a good trade-off since it's safer as the caller doesn't have to be particularly careful.

That said, I am okay with keeping the new logic, but maybe I'd add some documentation explicitly saying the operation is done in-place and as a result the input data can be altered. And maybe consider not returning anything so that's made even clearer.

Reply:
Funnily enough, we were doing `slice = model.RemoveDuplicateStrings(slice)` in a lot of places :laughing: